Feature: Add ordered waypoints

SuperCl4sh opened the pull request:
## What
Adds Coleweight's (and Soopy's?) ordered waypoints with /shordered or /sho. Should be compatible with Coleweight's routes.

## Changelog New Features
+ Added Coleweight's Ordered Waypoints (/shordered or /sho). - SuperClash

## Changelog Technical Details
+ Added AutoService to build.gradle.kts. - SuperClash

SuperCl4sh committed:
Add metal detector solver

SuperCl4sh committed:
Improve metal detector solver and add ordered waypoints (WIP)

SuperCl4sh committed:
Almost finish ordered waypoints

SuperCl4sh committed:
Fix ordered waypoints saving and importing

SuperCl4sh committed:
test

SuperCl4sh committed:
Make ordered waypoints actually work

SuperCl4sh committed:
Actually make it work

SuperCl4sh committed:
Merge branch 'hannibal002:beta' into coleweight

SuperCl4sh committed:
Finish ordered waypoints

SuperCl4sh committed:
Revert change

SuperCl4sh committed:
Hopefully fix detekt

SuperCl4sh committed:
Surely this time

SuperCl4sh committed:
Fix some bugs

SuperCl4sh committed:
Merge branch 'beta' into ordered-waypoints

SuperCl4sh committed:
Remove etherwarp

SuperCl4sh committed:
Fix detekt maybe and clean up code

SuperCl4sh committed:
Fix detekt last time

Jordyrat commented:
I've been using this, works well.

Only suggestion is for `/shoskip` to default to 1 with no args

SuperCl4sh committed:
Merge branch 'beta' into ordered-waypoints

SuperCl4sh committed:
Add default options for some commands, improve consistency of errors, and clean up code

SuperCl4sh committed:
Fix merge

SuperCl4sh committed:
Fix detekt

SuperCl4sh committed:
Merge branch 'beta' into ordered-waypoints

SuperCl4sh committed:
Maybe improve?

SuperCl4sh committed:
Finish first draft of ordered waypoints

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/SuperCl4sh/ordered-waypoints

# Conflicts:
#	build.gradle.kts

SuperCl4sh committed:
Merge branch 'beta' into ordered-waypoints

# Conflicts:
#	build.gradle.kts

SuperCl4sh committed:
Fix merge

SuperCl4sh committed:
Fix merge (2)

CalMWolfs committed:
make it build on 1.21

SuperCl4sh committed:
Add support for brigadier

SuperCl4sh committed:
Fix saving issues

SuperCl4sh committed:
Update OrderedWaypoints.kt

Fix detekt

SuperCl4sh committed:
Fix route saving by using more deep copies

SuperCl4sh committed:
Fix merge

SuperCl4sh committed:
Use periods instead of exclamation marks

SuperCl4sh committed:
Change some ChatUtils.chat to ChatUtils.userError, improve consistency, and clean up code

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/config/features/mining/orderedwaypoints/OrderedWaypointsConfig.kt:
```diff
@@ -0,0 +1,80 @@
+    @Expose
+    @ConfigOption(name = "Enable Ordered Waypoints", desc = "Enables ordered waypoints.")
+    @ConfigEditorBoolean
+    var enabled: Boolean = true
```
should have @FeatureToggle on this

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/config/features/mining/orderedwaypoints/OrderedWaypointsConfig.kt:
```diff
@@ -0,0 +1,80 @@
+    @Expose
+    @ConfigOption(name = "Next Color", desc = "Color of the next ordered waypoint.")
+    @ConfigEditorColour
+    var nextWaypointColor: String = "0:153:255:255:88"
```
Would be good if these (and the one or two further down) were initialised as a ChromaColour instead of as a string, theres a few examples of ChromaColour elsewhere in the config and HoppityCallWarningConfig has an example with rgba

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/config/storage/PlayerSpecificStorage.kt:
```diff
@@ -80,4 +82,7 @@ class PlayerSpecificStorage {
     }
+
+    @Expose
+    var routes: MutableMap<String, Waypoints<SkyhanniWaypoint>> = mutableMapOf()
```
i dont see a reason for these to be in player specific, id suggest making a new config file and storing them there.

CalMWolfs reviewed (changes requested):
havent looked too thoroughly through the code or tested at all yet, I should be able to do that tomorrow but I noticed these things that should all be changed

SuperCl4sh committed:
Add tab complete, move routes to the config file, add more options for ordered waypoints

SuperCl4sh committed:
Move the routes from OrderedWaypoints.kt to its own file

SuperCl4sh committed:
Fix detekt

SuperCl4sh committed:
Merge branch 'beta' into ordered-waypoints

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/SuperCl4sh/ordered-waypoints

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/SuperCl4sh/ordered-waypoints

CalMWolfs committed:
remove text, unused import, cleanup commands

CalMWolfs reviewed (changes requested):
I do not really ever mine but is the setup route line meant to be 2 blocks above the waypoint

<img width="879" height="1031" alt="image" src="https://github.com/user-attachments/assets/79cc7fa9-4626-4313-bbb1-480d98fcb0b7" />

I would also like there to be an option to instead of outlining the blocks you can also have it as a filled color instead

Also right now it only shows 1 waypoint ahead of where you are, i would like a slider that allows you to change the "look ahead" amount and probably default that to 2 and have it as a range of 1-5

also not really a problem but when i load the route it starts me at position 1 so it navigates me to position 2. I think best case would be it draws the line to the closest point in the route as you load it

I've also made a commit that that slightly changes around how the commands are registered so that its a bit easier to tab complete through them (especially on 1.21 as it shows usages there as you type)

the rest of it worked pretty well

SuperCl4sh committed:
Address comments and clean up code

SuperCl4sh commented:
> I do not really ever mine but is the setup route line meant to be 2 blocks above the waypoint
> 
> <img alt="image" width="879" height="1031" src="https://private-user-images.githubusercontent.com/94038482/466436956-79cc7fa9-4626-4313-bbb1-480d98fcb0b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI3MDg2MjEsIm5iZiI6MTc1MjcwODMyMSwicGF0aCI6Ii85NDAzODQ4Mi80NjY0MzY5NTYtNzljYzdmYTktNDYyNi00MzEzLWJiYjEtNDgwZDk4ZmNiMGI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE2VDIzMjUyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0ODVhNjBmMDI5OTY0NzBmODg1M2RiY2NiODIyNjZmMWFhZTMzMGY1ZWIxNWU1ZjFjMWEwMTRlZGYwNzk5ZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.E7U9w3dRfOenEBJqHhFjqwNkyJMZDh5c_9bkgSYhuzY">
> I would also like there to be an option to instead of outlining the blocks you can also have it as a filled color instead
> 
> Also right now it only shows 1 waypoint ahead of where you are, i would like a slider that allows you to change the "look ahead" amount and probably default that to 2 and have it as a range of 1-5
> 
> also not really a problem but when i load the route it starts me at position 1 so it navigates me to position 2. I think best case would be it draws the line to the closest point in the route as you load it
> 
> I've also made a commit that that slightly changes around how the commands are registered so that its a bit easier to tab complete through them (especially on 1.21 as it shows usages there as you type)
> 
> the rest of it worked pretty well

When using setup mode, the line is supposed to be the line from the user's eye when standing on the current waypoint to the next waypoint. It's to help clear the route if it's an etherwarp route. I've changed it to 1.62 blocks above the current waypoint if the user selects that they will not sneak (using conduit) and 1.54 if they will sneak (using aotv). I'm not sure how this works with 1.21. I've heard it's still 1.54 for sneaking height (Odin fabric uses this value), but I'm not sure.
I've added more config options (including one that lets the user choose how many waypoints ahead to render) and also changed the starting waypoint index to be whichever the user is closest to when they import the route. I've also done a bit of refactoring (namely for decideWaypoints), so it would be great if you could test it again. I've done some light testing and it worked fine.

SuperCl4sh committed:
Fix detekt and update outdated message

SuperCl4sh committed:
Merge branch 'hannibal002:beta' into ordered-waypoints

SuperCl4sh committed:
Remove unnecessary type adapter

SuperCl4sh committed:
Fix build

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/SuperCl4sh/ordered-waypoints

CalMWolfs committed:
default to 2

CalMWolfs committed:
dynamicSuggestionProvider function
